types/header.go

```diff
@@ -0,0 +1,229 @@
+		cdcEncode(h.Version),
+		cdcEncode(h.ChainID),
+		cdcEncode(h.Height),
+		cdcEncode(h.Time),
+		cdcEncode(h.LastBlockID),
+		cdcEncode(h.LastCommitHash),
+		cdcEncode(h.DataHash),
+		cdcEncode(h.ValidatorsHash),
+		cdcEncode(h.NextValidatorsHash),
+		cdcEncode(h.ConsensusHash),
+		cdcEncode(h.AppHash),
+		cdcEncode(h.LastResultsHash),
+		cdcEncode(h.EvidenceHash),
+		cdcEncode(h.ProposerAddress),
+	})
+}
+
+// StringIndented returns a string representation of the header
+func (h *Header) StringIndented(indent string) string {
+	if h == nil {
+		return "nil-Header"
+	}
+	return fmt.Sprintf(`Header{
+%s  Version:        %v
+%s  ChainID:        %v
+%s  Height:         %v
+%s  Time:           %v
+%s  LastBlockID:    %v
+%s  LastCommit:     %v
+%s  Data:           %v
+%s  Validators:     %v
+%s  NextValidators: %v
+%s  App:            %v
+%s  Consensus:      %v
+%s  Results:        %v
+%s  Evidence:       %v
+%s  Proposer:       %v
+%s}#%v`,
+		indent, h.Version,
+		indent, h.ChainID,
+		indent, h.Height,
+		indent, h.Time,
+		indent, h.LastBlockID,
+		indent, h.LastCommitHash,
+		indent, h.DataHash,
+		indent, h.ValidatorsHash,
+		indent, h.NextValidatorsHash,
+		indent, h.AppHash,
+		indent, h.ConsensusHash,
+		indent, h.LastResultsHash,
+		indent, h.EvidenceHash,
+		indent, h.ProposerAddress,
+		indent, h.Hash())
+}
+
+func (h Header) ValidateBasic() error {
+	if len(h.ChainID) > MaxChainIDLen {
+		return fmt.Errorf("chainID is too long. Max is %d, got %d", MaxChainIDLen, len(h.ChainID))
+	}
+
+	if h.Height < 0 {
+		return errors.New("negative Header.Height")
+	} else if h.Height == 0 {
+		return errors.New("zero Header.Height")
+	}
+
+	if err := h.LastBlockID.ValidateBasic(); err != nil {
+		return fmt.Errorf("wrong Header.LastBlockID: %v", err)
+	}
+	if err := ValidateHash(h.LastCommitHash); err != nil {
+		return fmt.Errorf("wrong Header.LastCommitHash: %v", err)
+	}
+
+	// Basic validation of hashes related to application data.
+	// Will validate fully against state in state#ValidateBlock.
+	if err := ValidateHash(h.ValidatorsHash); err != nil {
+		return fmt.Errorf("wrong Header.ValidatorsHash: %v", err)
+	}
+	if err := ValidateHash(h.NextValidatorsHash); err != nil {
+		return fmt.Errorf("wrong Header.NextValidatorsHash: %v", err)
+	}
+	if err := ValidateHash(h.ConsensusHash); err != nil {
+		return fmt.Errorf("wrong Header.ConsensusHash: %v", err)
+	}
+	// NOTE: AppHash is arbitrary length
+	if err := ValidateHash(h.LastResultsHash); err != nil {
+		return fmt.Errorf("wrong Header.LastResultsHash: %v", err)
+	}
+
+	if err := ValidateHash(h.EvidenceHash); err != nil {
+		return fmt.Errorf("wrong Header.EvidenceHash: %v", err)
+	}
+
+	if len(h.ProposerAddress) != crypto.AddressSize {
+		return fmt.Errorf("expected len(Header.ProposerAddress) to be %d, got %d",
+			crypto.AddressSize, len(h.ProposerAddress))
+	}
+
+	return nil
+}
+
+func (h *Header) ToProto() (*tmproto.Header, error) {
+	err := h.ValidateBasic()
```

Comment:
shouldn\'t we Validate in FromProto when we accept header from somewhere?

why do we need to Validate before marshalling?